src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -583,157 +542,259 @@ private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+                // event NOT firing because IsRequestInstrumented will return false without the headers.
+
                 var activity = new Activity(ActivityName);
+                activity.Start();
+
+                object asyncContext = s_readAResultAccessor(request);
+
+                // Step 1: Hook request._ReadAResult.m_AsyncState to store the state (request + callback + state + activity) we will need later.
+                s_asyncStateModifier(asyncContext, new Tuple<HttpWebRequest, object, AsyncCallback, Activity>(request, s_asyncStateAccessor(asyncContext), s_asyncCallbackAccessor(asyncContext), activity));
+
+                // Step 2: Hook request._ReadAResult.m_AsyncCallback so we can fire our events when the request is complete.
+                s_asyncCallbackModifier(asyncContext, s_asyncCallback);
+
+                InstrumentRequest(request, activity);
 
                 // Only send start event to users who subscribed for it, but start activity anyway
-                if (this.IsEnabled(RequestStartName))
+                if (IsEnabled(RequestStartName))
                 {
-                    this.StartActivity(activity, new { Request = request });
-                }
-                else
-                {
-                    activity.Start();
+                    Write(activity.OperationName + ".Start", new { Request = request });
                 }
+            }
+        }
 
-                if (activity.IdFormat == ActivityIdFormat.W3C)
-                {
-                    // do not inject header if it was injected already
-                    // perhaps tracing systems wants to override it
-                    if (request.Headers.Get(TraceParentHeaderName) == null)
-                    {
-                        request.Headers.Add(TraceParentHeaderName, activity.Id);
+        private void RaiseResponseEvent(HttpWebRequest request, HttpWebResponse response)
+        {
+            if (IsEnabled(RequestStopName))
+            {
+                Write(RequestStopName, new { Request = request, Response = response });
+            }
+        }
 
-                        var traceState = activity.TraceStateString;
-                        if (traceState != null)
-                        {
-                            request.Headers.Add(TraceStateHeaderName, traceState);
-                        }
-                    }
-                }
-                else
+        private void RaiseExceptionEvent(HttpWebRequest request, Exception exception)
+        {
+            if (IsEnabled(RequestExceptionName))
+            {
+                Write(RequestExceptionName, new { Request = request, Exception = exception });
+            }
+        }
+
+        private bool IsRequestInstrumented(HttpWebRequest request)
+        {
+            return request.Headers.Get(TraceParentHeaderName) != null || request.Headers.Get(RequestIdHeaderName) != null;
```

Comment:
should we check what mode we are operating in? It may allow less look ups in the collection.

Reply:
Done. Adding the check opened up a new avenue where you can have RequestId on there already but we'll still add TraceParent if the W3C mode is enabled so I added a test case for that.